src/hotspot/os/bsd/os_bsd.cpp

```diff
@@ -2039,13 +2039,28 @@ jint os::init_2(void) {
     if (status != 0) {
       log_info(os)("os::init_2 getrlimit failed: %s", os::strerror(errno));
     } else {
+#ifdef __APPLE__
+      rlim_t rlim_org = nbr_files.rlim_cur;
+      // according to setrlimit(2), OPEN_MAX must be used instead
+      // of RLIM_INFINITY, but testing on macOS >= 10.6, reveals that
+      // we can, in fact, use even RLIM_INFINITY, so try the max value
+      // that the system claims can be used first.
       nbr_files.rlim_cur = nbr_files.rlim_max;
 
-#ifdef __APPLE__
-      // Darwin returns RLIM_INFINITY for rlim_max, but fails with EINVAL if
-      // you attempt to use RLIM_INFINITY. As per setrlimit(2), OPEN_MAX must
-      // be used instead
-      nbr_files.rlim_cur = MIN(OPEN_MAX, nbr_files.rlim_cur);
+      // WORKAROUND: some terminals (ksh) will internally use "int" type
+      // to store this value and since RLIM_INFINITY overflows an "int"
+      // we might end up with a negative value, so cap the system limit max
+      // at INT_MAX instead, just in case, for everyone.
+      nbr_files.rlim_cur = MIN(INT_MAX, nbr_files.rlim_max);
+
+      status = setrlimit(RLIMIT_NOFILE, &nbr_files);
+      if (status != 0) {
```

Comment:
If status == 0 you will redundantly call setrlimit again at line 2066.

Reply:
Thank you for the feedback, fixed.

Since ksh exists on all the BSD platforms, not just Darwin, I changed the fix, so that the workaround is always active for all BSD OSes, which simplified the logic and let me fix the redundant call to setrlimit easily.